### PR TITLE
DF-1540: Send Mail to the bcc address

### PR DESCRIPTION
**Pull request by birom**

### CODE REVIEW

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java [hunk 1]
```diff
@@ -58,23 +62,41 @@ private Object sendMail(MailPreparator mailPreparator, Map<String, Object> heade
...
             }
         }
 
-        throw e;
+        return false;
+    }
+
+    private void handleException(MailPreparator mailPreparator) {
+        MailSenderData mailSenderData = mailPreparator.getMailSenderData();
+        MailSenderData fallbackSenderData = MailSenderData.builder()
```
Comment (pavax) [inline alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: we could create a method called copy(MailSenderData) on the MailSenderData.builder() and then just replace the setTo

```java
MailSenderData fallbackSenderData = MailSenderData.builder()
.copy(mailSenderData )
.setTo(this.mailSendingProperties.getBcc().toArray(new String[0]))
````

or


```java
MailSenderData fallbackSenderData = MailSenderData.builder(mailSenderData )
.setTo(this.mailSendingProperties.getBcc().toArray(new String[0]))
````

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java [hunk 2]
```diff
@@ -58,23 +62,41 @@ private Object sendMail(MailPreparator mailPreparator, Map<String, Object> heade
...
+                .setFrom(mailSenderData.getFrom())
+                .setSubject(mailSenderData.getSubject())
+                .setContent(mailSenderData.getContent())
+                .setTo(this.mailSendingProperties.getBcc().toArray(new String[0]))
+                .setCc(mailSenderData.getCc().toArray(new String[0]))
```
Comment (fabiangotzen) [inline alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: Not sure if its a valid concern, but shouldn't we ignore CC and BCC for the fallback?

For example, what would happen if we have an invalid mail in the CC of the complaintApplicationService?
```
   MailSenderData mailSenderData = new MailSenderData.Builder()
                .setTo(complaintProperties.getReceiverEmailAddress())
                .setCc(complaintDto.getContactInformation().getEmail())

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java [hunk 3]
```diff
@@ -33,12 +33,16 @@
...
     MailIntegrationFlowConfig(JavaMailSender mailSender,
             MailIntegrationChannels mailIntegrationChannels,
-            IntegrationBasisConfig integrationBasisConfig) {
+            IntegrationBasisConfig integrationBasisConfig,
+            MailSendingProperties mailSendingProperties) {
         this.mailSender = mailSender;
         this.inputChannel = mailIntegrationChannels.mailInputChannel();
         this.integrationBasisConfig = integrationBasisConfig;
+        this.mailSendingProperties = mailSendingProperties;
     }
```
Comment (hamidsoleymani) [inline alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: I would introduce the "final" key word. So we can see, that the reference is read-only and not reassigned.

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java [hunk 4]
```diff
@@ -58,23 +62,41 @@ private Object sendMail(MailPreparator mailPreparator, Map<String, Object> heade
...
-                return;
+                return true;
             }
         }
 
-        throw e;
+        return false;
+    }
+
+    private void handleException(MailPreparator mailPreparator) {
```
Comment (hamidsoleymani) [inline alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: 1)Maybe "handleException" should be called "resendEmailAfterException" 2) I don't know how often this method will be called. If frequently, creation of  contentPrefix should use StringBuilder and also should be in its own method, like createContentPrefix. Also creating of fallbackSender should be in its own method, like createFallbackSenderData(). If doing this way the method would be read:
resendEmailAfterException(){
createContentPrefix();
createFallbackSenderData();
this.mailSender.send(new MailPreparator(fallbackSenderData));
}

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java (hunk 2 above)
Comment (birom) [reply alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: fixed

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java (hunk 1 above)
Comment (birom) [reply alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: fixed

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java (hunk 3 above)
Comment (birom) [reply alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: Until now we have not been so strickt with _final_ keyword. This is consistent with the current code base. I would leave this for now  and discuss later if the whole team should change the current coding style.

File: alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java (hunk 4 above)
Comment (birom) [reply alv-mail-sending-starter/src/main/java/ch/admin/seco/alv/shared/mail/MailIntegrationFlowConfig.java]: It is called only in exceptional cases and the compiler takes care of this optimization in this case.

**Pull request merged**